### PR TITLE
Modularise/tidy/re-integrate krun_ext scripts.

### PR DESCRIPTION
**Pull request by vext01**
This does a fair amount of refactoring so that we can share code between
standalone running and the krun_ext scripts. Specifically:

 * Merge run_dacapo.py, run_spec.py and run_renaissance.sh into one
   Python script: run_standalone.py.

 * Get krun_ext_graal_ce.py, krun_ext_graal_ce_hotspot.py,
   krun_ext_openj9.py working again, sharing as much functionality with
   the standalone runner (via krun_ext_common.py).

 * Add a test target which does some quick checks.

Next step: rework Krun config file.

### CODE REVIEW

**Comment by vext01** [comment]
Additional:

I'm running the smoke test now. It will take a while to complete, hence the draft PR.

File: Makefile [hunk 1]
```diff
@@ -33,6 +33,12 @@ PATHS_SH=paths.sh
 
 TOP_DIR=`pwd`
 
+PEXECS=10
+INPROC_ITERS=2000
+
+SMOKE_PEXECS=1
+SMOKE_PEXECS=3
```
Comment (vext01) [inline Makefile]: Smoke test just crashed due to this ^ -- will fix.

File: krun_ext_common.py [hunk 2]
```diff
@@ -2,26 +2,233 @@
...
 import os
 
 
+PATHS = {}
+with open("paths.sh") as f:
+    for line in f:
+        k, v = line.strip().split("=")
+        PATHS[k] = v
+
+# Make sure these limits match those in the Krun config file.
```
Comment (ltratt) [inline krun_ext_common.py:19]: Could we extract them from the Krun config file perhaps?

File: krun_ext_common.py [hunk 3]
```diff
@@ -30,11 +237,82 @@ def run(args, benchmark, num_iters, tmp_file):
...
+        index = line.index("in")
+        assert line[index + 2] == "msec"
+        wallclock_times[idx] = float(Decimal(line[index + 1]) / 1000)
+        idx += 1
+
+    return wallclock_times
+
+
+def post_process_spec(done_exec):
+    # We have already check for non-zero exit status, but in addition SPEC
```
Comment (ltratt) [inline krun_ext_common.py]: "check" -> "checked".

Although this whole comment is quite lengthy and hard to parse. Can we tighten it up?

**Comment by ltratt** [comment]
For several reasons I'm not keen on the term "smoketest". Can we find something more descriptive like "quicktest" or similar?

File: krun_ext_common.py (hunk 2 above)
Comment (vext01) [reply krun_ext_common.py:19]: I'd intended on sharing lots of stuff, but the other way around: having the Krun config import this. However, I'm a little hesitant, as this is Python3 code, and Krun is Python2.

I bet it wouldn't be much effort to port Krun to Python3, but in the short term I think we can just duplicate.

File: krun_ext_common.py (hunk 3 above)
Comment (vext01) [reply krun_ext_common.py]: sure

**Comment by vext01** [comment]
> For several reasons I'm not keen on the term "smoketest"

We can.

But out of curiosity, why the objection? I learned this term from the Rust compiler guide, and before raising this PR I checked that it's  a generally accepted term:
https://en.wikipedia.org/wiki/Smoke_testing_(software)

**Comment by vext01** [comment]
Last push fixes everything but the duplication.

File: krun_ext_common.py [hunk 4]
```diff
@@ -267,9 +267,8 @@ def post_process_dacapo(done_exec):
 
 
 def post_process_spec(done_exec):
-    # We have already check for non-zero exit status, but in addition SPEC
-    # exits with a zero status when a benchmark (not the JVM) fails. Therefore,
-    # check stdout for errors.
+    # We have already checked for non-zero exit status, but SPEC exits zero
```
Comment (ltratt) [inline krun_ext_common.py]: How about "SPEC exits with a zero even if there were problems, so we need to check stdout in order to spot problems."?

Comment (vext01) [reply krun_ext_common.py]: We need to convey that there's no need to check the return code, as it was checked earlier.

Comment (ltratt) [reply krun_ext_common.py]: I don't see how that's relevant: whether we've checked it or not, by the time we've got to this point we've got check it anyway.

Comment (vext01) [reply krun_ext_common.py]: > whether we've checked it or not, by the time we've got to this point we've got check it anyway.

No, we don't have to check the return status, only stdout.

There seems to be confusion. We are checking two separate things:
 - The exit status (already checked in the callee)
 - stdout

In case a reader thinks that we have missed the former, I'd like the comment to be explicit.

Comment (ltratt) [reply krun_ext_common.py]: OK, I see. How about this: "Although at this point we've already checked that SPEC exited with zero, that isn't very helpful, because, surprisingly, SPEC always exits with zero, even when there are errors. We have to check stdout in order to spot errors."?

Also, I suspect we shouldn't be calling it "SPEC" but "SPECJVM" (or something like that)? SPEC is the organisation, not the benchmark suite itself.

Comment (vext01) [reply krun_ext_common.py]: > How about this

That works for me, but it's no shorter than before :P Maybe it's OK to be a long comment if it better describes the nuance?

> Also, I suspect we shouldn't be calling it "SPEC" but "SPECJVM"

Agreed.

Comment (ltratt) [reply krun_ext_common.py]: > Maybe it's OK to be a long comment if it better describes the nuance?

Yeah, it's not so much the length that's the killer, as that the shorter version was easily misread. Sometimes we can only achieve clarity by being a bit longer. Such is life :/

**Comment by vext01** [comment]
This should address all outstanding comments.

The official capitalisation the suite "SPECjvm", so I've used that where appropriate.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.
